File: LeetCode/0540. Single Element in a Sorted Array/Solution.kt
/**
 * Created by Inno Fang on 2018/2/25.
 */


/**
 * 7 / 7 test cases passed.
 * Status: Accepted
 * Runtime: 292 ms
 */
class Solution {
    fun singleNonDuplicate(nums: IntArray): Int {
        var lo = 0 ; var hi = nums.lastIndex / 2
        while (lo < hi) {
            val mid = lo + (hi - lo) / 2
            if (nums[mid * 2] != nums[mid * 2 + 1]) hi = mid
            else lo = mid + 1
        }
        return nums[lo * 2]
    }
}

fun main(args: Array<String>) {
    Solution().singleNonDuplicate(intArrayOf(1, 1, 2)).let(::println)
}

File: LeetCode/0540. Single Element in a Sorted Array/solution.cpp
/**
 * 15 / 15 test cases passed.
 * Runtime: 24 ms
 * Memory Usage: 21.9 MB 
 */
class Solution {
public:
    int singleNonDuplicate(vector<int>& nums) {
        int ans = 0;
        for (auto &num : nums) ans ^= num;
        return ans;
    }
};

/**
 * 15 / 15 test cases passed.
 * Runtime: 20 ms
 * Memory Usage: 21.8 MB 
 */
class Solution2 {
public:
    int singleNonDuplicate(vector<int>& nums) {
        int l = 0, r = nums.size() - 1;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (nums[mid] == nums[mid ^ 1]) l = mid + 1;
            else r = mid;
        }
        return nums[l];
    }
};
